light-table: add `iconSortable` property

ignatius-j opened the pull request:
Closes #463

buschtoens commented on addon/mixins/table-header.js:
```diff
@@ -84,9 +95,10 @@ export default Mixin.create({
+   * when the column is sorted descending.
    *
-   * @property iconDescending
+   * @property iconDesci
```
Whoops? :stuck_out_tongue:

buschtoens commented on addon/components/columns/base.js:
```diff
@@ -73,6 +73,21 @@ const Column = Component.extend(DraggableColumnMixin, {
+    }
+
+    return 'iconSortable';
+  }),
```
```js
  /**
   * @property sortIcon
   * @type {String|null}
   * @private
   */
  sortIcon: computed('column.{sortable,sorted,ascending}', function() {
    let sorted = this.get('column.sorted');

    if (sorted) {
      let ascending = this.get('column.ascending');
      return ascending ? 'iconAscending' : 'iconDescending';
    }

    let sortable = this.get('column.sortable');
    return sortable ? 'iconSortable' : null;
  }),
```

I think I would rather mark `sortIcon` as `private` (for now). And only return a `sortIcon`, if it was set. That way later down in the template, we can just call `{{if sortIcon}}` as to not insert a "meaningless" `<i>` tag.

This behavior is slightly different than the current one, but I think this makes more sense. I doubt that the "empty" `<i>` tag is used in the wild.

@alexander-alvarez What is your opinion?

alexander-alvarez replied:
Sounds good.
The only thing I can add is a :bike: :house_with_garden: on the name -- `sortIconProperty` or `sortIconAttribute` just so it's clear that this string represents the attribute that will be pulled from to generate the icon simply from the name (we could add docs for it too)

buschtoens commented on addon/templates/components/columns/base.hbs:
```diff
@@ -2,8 +2,8 @@
-  {{#if column.sorted}}
-    <i class="lt-sort-icon {{if column.ascending sortIcons.iconAscending sortIcons.iconDescending}}"></i>
+  {{#if column.sortable}}
+    <i class="lt-sort-icon {{get sortIcons sortIcon}}"></i>
```
With the other comment regarding `sortIcon` in mind, this would then be:

```hbs
{{#if sortIcon}}
   <i class="lt-sort-icon {{get sortIcons sortIcon}}"></i>
{{/if}}
```

ignatius-j replied:
👍 This all looks awesome!  I left a note about it in the issue, but I believe this implementation would still show:

```
<i class="lt-sort-icon"></i>
```

in the case that the user has set a column as sortable, but doesn't explicitly pass an `iconSortable` into their `t.head`.  Would we also need some sort of additional check to see if `iconSortable`/`iconAscending`/`iconDescending` is non-blank before showing the icon?  Or even just for `iconSortable`?  That way, the user has the option to not pass one in if they don't want to use it.  It may be ok with `<i class="lt-sort-icon">` showing for `iconAscending` or `iconDescending`, because those may be viewed as required fields to the component interface when dealing with columns that are `sortable`.

buschtoens commented:
Don't get thrown off by the *missing-tests* label. We didn't have any in the first place. :stuck_out_tongue_winking_eye: 

If you want, you can add some. If not, I'll do it afterwards. :wink:

ignatius-j commented:
@buschtoens  I noticed there weren't any tests explicitly for the sort icons.  I would be up for taking a crack at putting in some integration tests!

buschtoens commented:
I'd love that.

ignatius-j commented:
@buschtoens @alexander-alvarez I fixed the comments and added the initial integration test for making sure the sort icons render properly on `lt-head`.  We could split that test out into three tests (one for each sort icon) if you would like.  

Just let me know what you think!

alexander-alvarez commented on tests/integration/components/lt-head-test.js:
```diff
@@ -62,6 +63,35 @@ test('click - sortable column', function(assert) {
+  let sortIcon = find('.lt-sort-icon', sortableHeader);
+
+  // Sortable case
+  assert.ok(hasClass(sortIcon, 'fa-sort'), 'Sortable icon renders');
```
good test cases! I think it's fine as 1.

this also brought to light for me that the default behavior when sortable is enabled is to show the sortable icons.
If someone is expecting/want it not to show up it could be seen as a "breaking" change that the icon now show up.

Do we care? Should we document how turn it off? Have a cookbook example where it only shows on hover? Have a `showSortableIconByDefault` configuration option at the `ember-cli-build.js` level?

ignatius-j replied:
Before, if the column was sortable but hadn't been sorted yet, it would display the icon in the DOM but with just the `lt-sort-icon` class and no other classes (`<i class="lt-sort-icon"></i>`).

To get that same result now, you would just not pass in any `iconSortable` to `lt-head` (because by default `iconSortable` is `''`).  So it should probably be stated somewhere that it is an optional field that you shouldn't pass in if you don't want an icon for flagging that the column is sortable.

I could also try to put together a Cookbook example of an `iconSortable` that only shows on hover.  Let me know if you think that would be helpful!

alexander-alvarez replied:
woops :door: :walking_man:  didn't pay close enough attention to the default.
you're right.
